Integrate CMS for posts

markmhendrickson opened the pull request:
This PR implements a comprehensive CMS integration for the Leather web application, enabling the app to fetch, store, and display posts and corresponding content from the Leather CMS.

## Key Features
- Fetches content from the Leather CMS API at https://cms.leather.io/posts
- Stores posts as static JSON for optimized performance without runtime API calls
- Provides utility functions for accessing and formatting post content
- Adds TypeScript type definitions for post data structures
- Implements a content refresh mechanism via `fetch-posts.sh` script
- Integrates with existing pages to display CMS content (e.g., sBTC Rewards page)
- Adds a test page at `/test-posts` for verifying CMS integration

## Implementation Details
- Created data models and types for CMS content
- Added utility functions for working with post content
- Implemented string formatting utilities
- Set up a fetch script that can be run to update content
- Added a new npm script `fetch-posts` to package.json
- Integrated content in existing pages by replacing hardcoded strings with CMS content

## Testing
- Added a dedicated test page at `/test-posts` to demonstrate and verify the CMS integration
- Manually tested content rendering in existing pages
- Tested the fetch script to ensure it correctly retrieves and stores CMS data

## Next Steps
- Consider implementing automatic content fetching during build time
- Add more comprehensive error handling for failed fetches
- Implement caching strategies for improved performance

This integration makes the Leather website more maintainable by centralizing content management and allowing non-technical team members to update content without code changes.

## Screenshot example

<img width="1227" alt="image" src="https://github.com/user-attachments/assets/4e0815de-36ea-40e2-8dea-67126bf9465c" />

markmhendrickson commented:
cc @kyranjamie in case you want to review my approach here while still in flight. I'll proceed to integrate copy (e.g. section headers, hover cards) and content (e.g. FAQs) as represented by imported posts throughout the UI tomorrow.

camerow commented on mono.code-workspace:
```diff
@@ -0,0 +1,8 @@
+{
```
Remove this file

camerow commented on mono.code-workspace:
```diff
@@ -0,0 +1,8 @@
+{
```
Remove this file

markmhendrickson replied:
Removed

camerow commented on package.json:
```diff
@@ -221,5 +221,8 @@
       "pofile@1.1.4": "patches/pofile@1.1.4.patch",
       "expo-blur@13.0.2": "patches/expo-blur@13.0.2.patch"
     }
+  },
```
Remove this dependency

markmhendrickson replied:
Which one exactly?

camerow commented on apps/web/app/features/page/page.tsx:
```diff
@@ -17,8 +19,21 @@ export function PageInset(props: HTMLStyledProps<'div'>) {
+  postSlug?: string;
 }
-export function PageHeading({ title, subtitle }: PageHeadingProps) {
+export function PageHeading({ title, subtitle, disclaimer, postSlug }: PageHeadingProps) {
```
`Page` is incredibly generic language, if this is for the `explainers` I would call it such - definitely need something more specific than page.

markmhendrickson replied:
I've created a new `PostPageHeading` component for pages with headers that are represented by posts, and it accepts just a post parameter.

camerow commented on apps/web/app/pages/sbtc-rewards/sbtc-rewards.tsx:
```diff
@@ -40,7 +42,7 @@ const sbtcEnroll = {
   logo: <SbtcLogo size="32px" />,
   title: 'Basic sBTC rewards',
-  description: 'Hold sBTC in your wallet to passively earn more sBTC, as it compounds over time.',
+  description: formatPostPrompt(content.posts["sbtc-rewards-basic"]?.Prompt || ''),
```
Fallback to the original text?

markmhendrickson replied:
I spoke with Kyran about this briefly on Friday. I believe @kyranjamie and I were both thinking that if content is loaded at _build_ time (rather than runtime) then fallbacks shouldn't be needed. 

My main concern with fallbacks in the code (for copy and content generally) is that they can easily get out of date with the source of truth, which may be more dangerous than simply throwing a build-time error that prompts source correction. But perhaps worth a discussion?

camerow commented on apps/web/app/pages/sbtc-rewards/sbtc-rewards.tsx:
```diff
@@ -135,8 +138,10 @@ export function SbtcRewards() {
         <Page.Heading
-          title="Earn yield with Bitcoin on Stacks"
-          subtitle="Get BTC in the form of sBTC on Stacks—the leading L2 for Bitcoin—to earn yield from a variety of reward protocols without sacrificing Bitcoin’s underlying security and self-sovereignty."
+          title={formatPostPrompt(content.posts[postSlug]?.Prompt || '')}
```
`Page` should just accept a `postSlug` and read `content` internally, rather than both taking a slug and all of the extra props.

camerow commented on apps/web/app/utils/post-utils.ts:
```diff
@@ -0,0 +1,28 @@
+ * @param prompt The prompt content from a post
+ * @returns The formatted prompt
+ */
+export function formatPostPrompt(prompt: string | undefined | null): string {
```
These functions all do exactly the same thing, `removeTrailingPeriod` - what's the point of them?

markmhendrickson replied:
I'll consolidate in the next commit

camerow commented on apps/web/app/utils/posts.ts:
```diff
@@ -0,0 +1,37 @@
+ * Get featured posts
+ */
+export function getFeaturedPosts(): Post[] {
+  return getAllPosts().filter(post => post.Featured);
```
```suggestion
  return getAllPosts().filter(post => !!post.Featured);
```

camerow commented on apps/web/app/scripts/fetch-posts.js:
```diff
@@ -0,0 +1,97 @@
+/**
+ * Fetches all posts from the CMS by iterating through all pages
+ */
+async function fetchAllPosts() {
```
These scripts should probably live a directory up since they are not part of `app`

markmhendrickson replied:
I'll move in next commit

camerow commented on apps/web/app/scripts/fetch-posts.js:
```diff
@@ -0,0 +1,97 @@
+  while (hasNextPage) {
+    try {
+      console.log(`Fetching page ${currentPage}...`);
+      const response = await fetch(`https://cms.leather.io/posts?page=${currentPage}`);
```
Why paginate the requests if we plan to use this to generate a `json` file with all f the responses? I think better to create an endpoint with one response for everything.

markmhendrickson replied:
The request is no longer paginated in latest commit

camerow commented on apps/web/app/scripts/fetch-posts.js:
```diff
@@ -0,0 +1,97 @@
+  
+  posts.forEach(post => {
+    // Skip posts without a slug
+    if (!post.Slug) {
```
Something that should be enforced at the API layer I think.

markmhendrickson replied:
This is resolved in latest commit

camerow commented on apps/web/app/scripts/fetch-posts.js:
```diff
@@ -0,0 +1,97 @@
+    }
+    
+    // If we already have a post with this slug, only keep the most recently created one
+    if (postsMap[post.Slug]) {
```
The capitalized json properties are throwing me off, it's a standard across our codebase to use lowercase for `json` properties.

markmhendrickson replied:
I'll see about converting to lowercase upon import of content

kyranjamie reviewed (changes requested):
Left a handful of comments, but more broadly I'm concerned this work creates too tight of a dependency on the CMS.

I'd be much more comfortable with some sort of architecture where, for example, the CMS synced to a Github repo that we reference locally at dev-time & clone at build-time. Not a very common set up, but this way if the CMS goes down, the web app continues to build with the most recent data. This repo could enforce it's own schema definitions and validations etc with CI. It would allow the CMS to be entirely refactored / changed, without requiring changes here.

markmhendrickson commented:
<!-- linear:replyTo:2845653968 -->

> this way if the CMS goes down, the web app continues to build with the most recent data. This repo could enforce it's own schema definitions and validations etc with CI

Isn't this the case with this current approach though? In that, the import script creates a copy of the most recent data from the CMS but then there are no further dependencies on the CMS. This approach also involves enforcing its own schema definitions upon import.

markmhendrickson commented:
<!-- linear:replyTo:2845653968 -->

The script is also referenced and executed only at dev-time and not in production / triggered by usage per se.

markmhendrickson commented:
### Summary of Changes

- **Content-driven UI for sBTC and Stacking pages:**  
  - Refactored sBTC Rewards and Stacking pages/components to use dynamic content from `content.posts` and new utility functions for post links and sanitization.
  - Added new components: `PostPageHeading`, `PostSectionHeading`, `PostLabelHoverCard`, and `PostValueHoverCard` for rendering content-driven headings, labels, and value displays.
  - Updated FAQ and explainer components (`SbtcRewardsFaq`, `StackingFaq`, `LiquidStackingExplainer`, `StackingExplainer`) to render questions and answers from content, with "Learn more" links.
  - Replaced hardcoded text in value cards, tables, and protocol grids with content-driven labels and tooltips.
  - Improved accessibility and test coverage with new and updated test files for all new/changed components.

- **Component and Data Structure Updates:**  
  - Removed deprecated and unused components and scripts (e.g., `stacking-instructions.tsx`, `apy-hero-card.tsx`, `fetch-posts` scripts).
  - Updated and extended types and data structures in `post-types.ts`, `content.ts`, and related files.
  - Added new utility functions for content sanitization and post linking.

- **UI/UX Improvements:**  
  - Improved label consistency and dynamic rendering in stacking forms, protocol tables, and reward grids.
  - Updated footer links and removed the static disclaimer in favor of content-driven legal text.
  - Enhanced test coverage for XSS protection and content-driven UI.

- **Other:**  
  - Updated and cleaned up various package and config files.
  - Added/updated documentation in `README.md` files.

### New Post Fetch Mechanism on Build

A new mechanism has been introduced to fetch and integrate post content at build time, ensuring that dynamic content (such as FAQs, explainers, and UI labels) is always up-to-date and centrally managed.

**How it works:**
- A new script (`scripts/fetch-remote-posts.cjs`) is executed during the build process.
- This script fetches post data from a remote source (e.g., a CMS, API, or static endpoint) and generates the necessary data files in the codebase (namely `apps/web/app/data/posts.json`).
- The fetched content is then imported and used throughout the app for rendering dynamic UI elements, including headings, tooltips, FAQs, and more.
- This approach eliminates the need for manual updates to static content in the codebase, reduces duplication, and ensures consistency across the UI.

**Benefits:**
- **Centralized content management:** All post-related content is managed from a single source of truth.
- **Automatic updates:** Content changes are reflected in the UI on the next build without manual intervention.
- **Improved maintainability:** Reduces the risk of stale or inconsistent content across different parts of the app.
- **Enhanced flexibility:** Enables non-developers to update UI content via the remote source, streamlining the content update workflow.

kyranjamie commented on apps/web/app/components/post-label-hover-card.test.tsx:
```diff
@@ -0,0 +1,31 @@
+// See: https://reactjs.org/link/invalid-hook-call
+// Re-enable after standardizing React/ReactDOM versions across all packages.
+
+import { render, screen, fireEvent } from '@testing-library/react';
```
Please don't use react testing library, we can use playwright for this equally as well

kyranjamie commented on apps/web/app/features/stacking/start-pooled-stacking/components/choose-pooling-duration.tsx:
```diff
@@ -9,8 +9,8 @@ export function ChoosePoolingDuration() {
           <styled.p textStyle="caption.01">
-            The pool stacks your STX for up to 12 cycles. You can revoke anytime, but they stay
-            locked until the cycle ends. Revoke before restacking to regain access.
+            The pool will commit your STX for Stacking for up to 12 cycles (with about two weeks per cycle). You can revoke anytime, but they stay
```
passive voice

markmhendrickson replied:
Because it's in future tense, suggesting this change?

"The pool commits your STX to Stacking for up to 12 cycles. You can revoke before a cycle restarts to regain access."

kyranjamie commented on apps/web/app/components/pool-overview.tsx:
```diff
@@ -113,26 +149,91 @@ export function PoolOverview({ pool, poolSlug }: PoolOverviewProps) {
 }
+
+// Add a component for the stacking-amount post label
+export function StackingAmountLabel({ textStyle = "label.03", tagName = "h1" }: { textStyle?: string; tagName?: keyof JSX.IntrinsicElements }) {
+  const label = (content.posts as Record<string, any>)["stacking-amount"]?.Title || "Amount";
```
this is very messy code

markmhendrickson replied:
How so?

fbwoolf replied:
Typ we would so something like...

```ts
const label = 'stacking-amount' in content.posts ? content.posts.title : 'Amount'
```

Ideally we always lowercase props like `title` (at least we do in the extension).

kyranjamie commented on apps/web/app/components/post-info-hover-icon.tsx:
```diff
@@ -0,0 +1,45 @@
+export function PostInfoHoverIcon({ postKey, children, iconColor = 'black' }: PostInfoHoverIconProps) {
+  const post = (content.posts as Record<string, any>)[postKey];
+  if (!post) return <>{children}</>;
+  const iconColorToken = iconColor === 'white' ? 'invert' : 'ink.text-subdued';
```
I'm pretty sure this won't work in Panda https://panda-css.com/docs/guides/dynamic-styling

markmhendrickson replied:
It seems to work here?

<img width="294" alt="image" src="https://github.com/user-attachments/assets/60cb303b-9734-465b-85f8-dee271de96bf" />

kyranjamie commented on apps/web/app/components/post-info-hover-icon.tsx:
```diff
@@ -0,0 +1,45 @@
+  const iconColorToken = iconColor === 'white' ? 'invert' : 'ink.text-subdued';
+  const handleIconClick = (e: React.MouseEvent) => {
+    e.stopPropagation();
+    window.location.assign(getPostHref(post.Slug));
```
This should use react router

markmhendrickson replied:
I'll change in next commit

markmhendrickson replied:
Resolving in next push

markmhendrickson commented on apps/web/app/data/post-types.ts:
```diff
@@ -0,0 +1,37 @@
+  Hidden: boolean;
+  Question: string;
+  Prompt: string;
+  Images: Array<{ name: string; url: string }>;
```
Which convention(s) in particular?

fbwoolf replied:
I see this now, one thing is to lowercase, and camelCase when needed ...not hyphenated. Hyphenated is typ only seen when a type coming from an api, we then convert it to our own types using camelCase which is more common in frontend dev.

markmhendrickson replied:
Resolved in next push

fbwoolf commented on apps/web/app/components/post-page-heading.test.tsx:
```diff
@@ -0,0 +1,54 @@
+
+const mockPost = {
+  id: '1',
+  Title: 'Test Title',
```
Can we lowercase these and is there a type for the `Post` to use throughout the codebase?

markmhendrickson replied:
[Related](https://github.com/leather-io/mono/pull/1193#discussion_r2095738095)

markmhendrickson commented on apps/web/app/routes/test-posts.tsx:
```diff
@@ -0,0 +1,44 @@
+        <h2 className="text-xl font-semibold mb-2">Post Statistics</h2>
+        <p>Total posts: {allPosts.length}</p>
+        <p>Explainers category posts: {explainers.length}</p>
+      </div>
```
I've removed this route so it should be resolved implicitly.

fbwoolf commented on apps/web/app/features/stacking/user-positions.tsx:
```diff
@@ -159,14 +161,22 @@ export function UserPositions({ stacksAddress }: UserPositionsProps) {
         <InfoGrid.Cell>
-          <ValueDisplayerWithLoader name="APR" isLoading={fakeLoading} value="~10%" />
+          <ValueDisplayerWithLoader
+            name={<PostLabelHoverCard postKey="historical-yield" label={(content.posts as Record<string, any>)["historical-yield"]?.Title || "APR"} textStyle="label.03" />}
```
If these are properly typed there shouldn't be a need to use the `as` keyword to cast them.

markmhendrickson replied:
Resolved in next push

markmhendrickson commented:
@fbwoolf good question. @kyranjamie can we get a PR-based staging environment in place for the web app inside the mono repo?

kyranjamie commented:
https://linear.app/leather-io/issue/LEA-2640/pr-demo-builds-for-web-app

camerow commented on apps/web/app/components/explainer.tsx:
```diff
@@ -66,11 +70,11 @@ export function ExplainerStep({ index, title, description, img }: EarnInstructio
         </VStack>
         <styled.h3 mt="space.02" textStyle="label.01">
-          {title}
+          {post ? <PostLabelHoverCard postKey={post.Slug} label={String(title)} /> : title}
```
This seems like a really flaky contract - sometimes title is a string but other times it's a `ReactNode`? When is it a string and when is it a node and why?

markmhendrickson commented:
**Summary of All Changes for latest PR Commit:**

- **Centralized Content Management:**  
  - Migrated all hardcoded UI copy (labels, descriptions, pool/protocol data, stacking/liquid stacking/pooled stacking conditions, status messages, section headers, etc.) into `apps/web/app/data/content.ts` for consolidated management.
  - Updated all relevant components to reference this centralized content, removing hardcoded strings.
  - Used backticks for all `description` fields in `content.ts` for consistency and future flexibility.

- **Component Refactors:**  
  - Refactored sBTC rewards, stacking provider tables, stacking/liquid stacking/pooled stacking conditions, and stacking info components to use the new content structure.
  - Updated icon and logo mapping to use string keys from content.
  - Updated table headers, button labels, and status messages to use content-driven values.

- **File Cleanup:**  
  - Deleted redundant or now-unused utility and test files (`post-hooks.ts`, `posts.ts`, old test files, etc.).
  - Removed hardcoded arrays and strings from components, replacing with content references.

- **Test and Config Updates:**  
  - Added/updated E2E tests for post-related components.
  - Updated Playwright config and lockfiles as needed.

- **Minor Content Edits:**  
  - Adjusted some explainer step titles in `content.ts` for clarity and consistency.

---

**Comprehensive Commit Message (≤100 characters):**

Centralize UI copy in content.ts, refactor components, remove hardcoded strings, clean up files

markmhendrickson commented on apps/web/app/data/README.md:
```diff
@@ -0,0 +1,79 @@
+  Featured: boolean;
+  Hidden: boolean;
+  // ...and more fields
+}
```
Resolved in next push

markmhendrickson commented on packages/utils/src/sanitize-content.spec.ts:
```diff
@@ -0,0 +1,28 @@
+  it('removes <script> tags in browser', () => {
+    // Simulate browser environment
+    (global as any).window = {};
+    (global as any).DOMPurify = { sanitize: (dirty: string) => dirty.replace(/<script.*?>.*?<\/script>/gi, '') };
```
Resolved in next push

markmhendrickson commented on packages/utils/src/sanitize-content.spec.ts:
```diff
@@ -0,0 +1,28 @@
+  it('removes <script> tags in browser', () => {
+    // Simulate browser environment
+    (global as any).window = {};
+    (global as any).DOMPurify = { sanitize: (dirty: string) => dirty.replace(/<script.*?>.*?<\/script>/gi, '') };
```
Resolved in next push
